llvm/cmake/modules/AddSecurityFlags.cmake

```diff
@@ -31,73 +31,199 @@ macro(add_link_option_ext flag name)
+     OR (is_icpx AND MSVC))
+    add_compile_option_ext("-fPIC" FPIC)
+  elseif(is_msvc)
+    add_compile_option_ext("/Gy" GY)
+  endif()
+
+  # Position Independent Execution
+  if(is_gcc
+     OR is_clang
+     OR (is_icpx AND MSVC))
+    # The project should be configured with -DCMAKE_POSITION_INDEPENDENT_CODE=ON
+    add_compile_option_ext("-fPIE" FPIE)
+    add_link_option_ext("-pie" PIE CMAKE_EXE_LINKER_FLAGS
+                        CMAKE_MODULE_LINKER_FLAGS CMAKE_SHARED_LINKER_FLAGS)
+  elseif(is_msvc)
+    add_compile_option_ext("/DYNAMICBASE" DYNAMICBASE)
+  endif()
+
+  if(CMAKE_BUILD_TYPE MATCHES "Release")
+    if(is_msvc)
+      add_compile_option_ext("/NXCOMPAT" NXCOMPAT)
+    endif()
+  endif()
+
+  # Stack Protection
+  if(is_msvc)
+    add_compile_option_ext("/GS" GS)
+  elseif(
+    is_gcc
+    OR is_clang
+    OR (is_icpx AND MSVC))
+    if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+      add_compile_option_ext("-fstack-protector" FSTACKPROTECTOR)
+    elseif(CMAKE_BUILD_TYPE MATCHES "Release")
+      add_compile_option_ext("-fstack-protector-strong" FSTACKPROTECTORSTRONG)
+      add_compile_option_ext("-fstack-clash-protection" FSTACKCLASHPROTECTION)
+    endif()
+  endif()
+
+  if(LLVM_ON_UNIX)
     # Fortify Source (strongly recommended):
-    if (CMAKE_BUILD_TYPE STREQUAL "Debug")
-      message(WARNING
-        "-D_FORTIFY_SOURCE=2 can only be used with optimization.")
-      message(WARNING "-D_FORTIFY_SOURCE=2 is not supported.")
+    if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+      message(WARNING "-D_FORTIFY_SOURCE=3 can only be used with optimization.")
+      message(WARNING "-D_FORTIFY_SOURCE=3 is not supported.")
     else()
-      # Sanitizers do not work with checked memory functions,
-      # such as __memset_chk. We do not build release packages
-      # with sanitizers, so just avoid -D_FORTIFY_SOURCE=2
-      # under LLVM_USE_SANITIZER.
-      if (NOT LLVM_USE_SANITIZER)
-        message(STATUS "Building with -D_FORTIFY_SOURCE=2")
-        add_definitions(-D_FORTIFY_SOURCE=2)
+      # Sanitizers do not work with checked memory functions, such as
+      # __memset_chk. We do not build release packages with sanitizers, so just
+      # avoid -D_FORTIFY_SOURCE=3 under LLVM_USE_SANITIZER.
+      if(NOT LLVM_USE_SANITIZER)
+        message(STATUS "Building with -D_FORTIFY_SOURCE=3")
+        add_definitions(-D_FORTIFY_SOURCE=3)
       else()
-        message(WARNING
-          "-D_FORTIFY_SOURCE=2 dropped due to LLVM_USE_SANITIZER.")
+        message(
+          WARNING "-D_FORTIFY_SOURCE=3 dropped due to LLVM_USE_SANITIZER.")
       endif()
     endif()
 
-    # Format String Defense
-    add_compile_option_ext("-Wformat" WFORMAT)
-    add_compile_option_ext("-Wformat-security" WFORMATSECURITY)
-    add_compile_option_ext("-Werror=format-security" WERRORFORMATSECURITY)
-
-    # Stack Protection
-    add_compile_option_ext("-fstack-protector-strong" FSTACKPROTECTORSTRONG)
+    add_definitions(-D_GLIBCXX_ASSERTIONS)
```

Comment:
I understand that our guideline says to set it for both release and debug modes, but I would prefer to avoid it having set for builds that we will publish in our releases.

The reason for this is that extra checks impact performance and whilst its good to have them in our testing, I don't think that their presence is justified for our production release builds.

Let's maybe tie it to assertions being enabled? That would make the most sense to me